### PR TITLE
Conditional upserts for 8.6.2

### PR DESCRIPTION
**Pull request by johnmbw**
Basically the changes from:

* https://github.com/BrandwatchLtd/bw-lucene-solr/pull/47
* https://github.com/BrandwatchLtd/bw-lucene-solr/pull/49

but cherry-picked to the 8.6.2 branch.

Had to add one extra commit to fix some compilation issues too.

### CODE REVIEW

**Comment by johnmbw** [comment]
I need to probably do some manual testing of this first.

File: solr/core/src/java/org/apache/solr/update/processor/ConditionalUpsertProcessorFactory.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        return false;
+      }
+      return distribProc.isLeader();
+    }
+
+    @Override
+    public void processAdd(AddUpdateCommand cmd) throws IOException {
+      if (!ignoreConditionalUpserts && !conditions.isEmpty() && isLeader(cmd)) {
+        BytesRef indexedDocId = cmd.getIndexedId();
+        SolrInputDocument oldDoc = RealTimeGetComponent.getInputDocument(core, indexedDocId, RealTimeGetComponent.Resolution.DOC);
```
Comment (richardbadman) [inline solr/core/src/java/org/apache/solr/update/processor/ConditionalUpsertProcessorFactory.java:126]: I wonder if [this commit Tim did](https://github.com/BrandwatchLtd/bw-lucene-solr/commit/eccb3e52961698db1e3b4778810a87003d0349ee) would need to be applied here? Without opening this in intellij / looking at the java docs for the `getInputDocument` I don't know for sure, but Tim did have to add some extra parameters for the 8.11.2 branch with the commit being:
> Fixes for solr8 build

**Review by richardbadman — COMMENTED** [review]
Appreciate this is a draft PR still, but saw the following and thought I'd mention it just incase it wasn't on your radar already

File: solr/core/src/java/org/apache/solr/update/processor/ConditionalUpsertProcessorFactory.java (hunk 1 above)
Comment (johnmbw) [reply solr/core/src/java/org/apache/solr/update/processor/ConditionalUpsertProcessorFactory.java:126]: yeah this was essentially the difference I was talking about.  I'd just chose a difference method for my fix.

**Comment by richardbadman** [comment]
I took a look at the method and it is different in 8.11.2 like you thought @johnmbw , but the tests pass and so only a small minor change added from the patch Tim added to the 8.11.2 branch anyway ,

**Review by richardbadman — APPROVED** [review]

**Pull request merged**